xdsl/transforms/lower_riscv_func.py

```diff
@@ -59,9 +59,84 @@ def match_and_rewrite(self, op: riscv_func.SyscallOp, rewriter: PatternRewriter)
         rewriter.replace_matched_op(ops, new_results=new_results)
 
 
+SCALL_EXIT = 93
+"""
+93 is the number of the `exit` syscall on RISCV.
+"""
+
+
+class LowerRISCVFuncOp(RewritePattern):
+    """
+    Temporary lowering of only the main function, would like to lower non-main functions
+    soon.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv_func.FuncOp, rewriter: PatternRewriter):
+        body = op.func_body.block
+        first_op = body.first_op
+        assert first_op is not None
+        while len(body.args):
+            # arguments are passed to riscv functions via a0, a1, ...
+            # replace arguments with `GetRegisterOp`s
+            index = len(body.args) - 1
+            last_arg = body.args[-1]
+            get_reg_op = riscv.GetRegisterOp(riscv.Register(f"a{index}"))
+            last_arg.replace_by(get_reg_op.res)
+            rewriter.insert_op_before(get_reg_op, first_op)
+            first_op = get_reg_op
+            rewriter.erase_block_argument(last_arg)
+
+        rewriter.inline_block_after(op.func_body.block, op)
+        rewriter.replace_matched_op(riscv.LabelOp(op.func_name.data))
+
+
+class LowerRISCVFuncReturnOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv_func.ReturnOp, rewriter: PatternRewriter):
+        func_op = op.parent_op()
+        if func_op is not None:
+            assert isinstance(func_op, riscv_func.FuncOp)
+            if func_op.func_name.data == "main":
+                # Main does not return, call exit syscall instead
```

Comment:
Where would it return to? My understanding is that `main` is kind of just the entry point to the `.text` section, and not a "real" function

Reply:
`main` always returns as a regular function, what happens later depends on the platform. On bare metal it's usually a super simple wrapper (`crt0`) that takes care of the return value but since we already have syscalls in the mix we are assuming a full fledged OS (e.g.: on `linux-gnu` is `_libc_start_main` that also does some flavor of `exit` syscall). I would say that the compiler has zero knowledge of who's going to invoke `main` (not really sure it's always true tho...).

Reply:
...it's a much bigger topic tho, maybe this PR is not the right place to figure it out :) Approving as it is because it's really nice overall.

Reply:
Ah no I'd love your input, we don't need to merge it as-is. It seems that we could simplify this PR by removing the syscall special case, if I understand you correctly. At the least, some lever higher than riscv-func could be made to deal with this, maybe the llvm to riscv lowering.

Reply:
> Maybe the best would be to get rid of the special case in this PR, and take some time later to decide how we want to handle main

Absolutely agree, removing any special case about `main` would be both easier and cleaner.

Reply:
No objection there.

Reply:
Thank you for the discussion, removed special handling of `main` function

Reply:
Thanks for the changes, looks super neat!